contracts/orderbook/Logic/OrderBook.sol

```diff
@@ -0,0 +1,413 @@
+                    order.base,
+                    order.loanTokenAddress,
+                    order.collateralTokenAmount
+                );
+                amountUsed = (amountUsed * order.leverage) / 10**18;
+                (dSwapValue, ) = swapImpl.dexAmountOutFormatted(
+                    payload,
+                    amountUsed
+                );
+                dSwapValue += order.collateralTokenAmount;
+            }
+
+
+            if (order.amountReceived <= dSwapValue) {
+                return true;
+            }
+        } else if (order.orderType == IOrderBook.OrderType.LIMIT_CLOSE) {
+            if (!_isActiveLoan(order.loanID)) {
+                return false;
+            }
+            uint256 dSwapValue;
+            if (order.isCollateral) {
+                (dSwapValue, ) = swapImpl.dexAmountInFormatted(
+                    payload,
+                    order.collateralTokenAmount
+                );
+            } else {
+                (dSwapValue, ) = swapImpl.dexAmountOutFormatted(
+                    payload,
+                    order.collateralTokenAmount
+                );
+            }
+            if (order.amountReceived <= dSwapValue) {
+                return true;
+            }
+        } else {
+            if (!_isActiveLoan(order.loanID)) {
+                return false;
+            }
+            bool operand;
+            if (_useOracle[order.trader]) {
+                operand =
+                    order.amountReceived >=
+                    queryRateReturn(
+                        order.base,
+                        order.loanTokenAddress,
+                        order.collateralTokenAmount
+                    );
+            } else {
+                operand =
+                    order.amountReceived >=
+                    getDexRate(
+                        swapImpl,
+                        order.base,
+                        order.loanTokenAddress,
+                        payload,
+                        order.collateralTokenAmount
+                    );
+            }
+            if (operand) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    function getDexRate(
+        ISwapsImpl swapImpl,
+        address base,
+        address loanTokenAddress,
+        bytes memory payload,
+        uint256 amountIn
+    ) public returns (uint256 rate) {
+        (rate, ) = swapImpl.dexAmountOutFormatted(
+            payload,
+            10**IERC20Metadata(base).decimals()
+        );
+        rate = (rate * amountIn) / 10**IERC20Metadata(base).decimals();
+    }
+
+    function priceCheck(
+        address loanTokenAddress,
+        address base,
+        ISwapsImpl swapImpl,
+        bytes memory payload
+    ) public returns (bool) {
+        uint256 dexRate = getDexRate(
+            swapImpl,
+            base,
+            loanTokenAddress,
+            payload,
+            10**IERC20Metadata(base).decimals()
```

Comment:
this decimal() call can be cashed in local storage

Reply:
will change